src/nvim/quickfix.c

```diff
@@ -428,6 +446,505 @@ static efm_T * parse_efm_option(char_u *efm)
+  // Get the next line from the supplied list
+  while (p_li != NULL
+         && (p_li->li_tv.v_type != VAR_STRING
+             || p_li->li_tv.vval.v_string == NULL)) {
+    p_li = p_li->li_next;               // Skip non-string items
+  }
+
+  if (p_li == NULL) {                   // End of the list
+    state->p_li = NULL;
+    return QF_END_OF_INPUT;
+  }
+
+  len = STRLEN(p_li->li_tv.vval.v_string);
+  if (len > IOSIZE - 2) {
+    state->linebuf = qf_grow_linebuf(state, len);
+  } else {
+    state->linebuf = IObuff;
+    state->linelen = len;
+  }
+
+  STRLCPY(state->linebuf, p_li->li_tv.vval.v_string, state->linelen + 1);
+
+  state->p_li = p_li->li_next;                 // next item
+  return QF_OK;
+}
+
+/// Get the next string from state->buf.
+static int qf_get_next_buf_line(qfstate_T *state)
+{
+  char_u *p_buf = NULL;
+  size_t len;
+
+  // Get the next line from the supplied buffer
+  if (state->buflnum > state->lnumlast) {
+    return QF_END_OF_INPUT;
+  }
+  p_buf = ml_get_buf(state->buf, state->buflnum, false);
+  state->buflnum += 1;
+
+  len = STRLEN(p_buf);
+  if (len > IOSIZE - 2) {
+    state->linebuf = qf_grow_linebuf(state, len);
+  } else {
+    state->linebuf = IObuff;
+    state->linelen = len;
+  }
+  STRLCPY(state->linebuf, p_buf, state->linelen + 1);
+
+  return QF_OK;
+}
+
+/// Get the next string from file state->fd.
+static int qf_get_next_file_line(qfstate_T *state)
+{
+  size_t growbuflen;
+
+  if (fgets((char *)IObuff, IOSIZE, state->fd) == NULL) {
+    return QF_END_OF_INPUT;
+  }
+
+  bool discard = false;
+  state->linelen = STRLEN(IObuff);
+  if (state->linelen == IOSIZE - 1 && !(IObuff[state->linelen - 1] == '\n'
+#ifdef USE_CRNL
+                                        || IObuff[state->linelen - 1] == '\r'
+#endif
+                                        )) {  // NOLINT(whitespace/parens)
+    // The current line exceeds IObuff, continue reading using growbuf
+    // until EOL or LINE_MAXLEN bytes is read.
+    if (state->growbuf == NULL) {
+      state->growbufsiz = 2 * (IOSIZE - 1);
+      state->growbuf = xmalloc(state->growbufsiz);
+    }
+
+    // Copy the read part of the line, excluding null-terminator
+    memcpy(state->growbuf, IObuff, IOSIZE - 1);
+    growbuflen = state->linelen;
+
+    for (;;) {
+      if (fgets((char *)state->growbuf + growbuflen,
+                (int)(state->growbufsiz - growbuflen), state->fd) == NULL) {
+        break;
+      }
+      state->linelen = STRLEN(state->growbuf + growbuflen);
+      growbuflen += state->linelen;
+      if (state->growbuf[growbuflen - 1] == '\n'
+#ifdef USE_CRNL
+          || state->growbuf[growbuflen - 1] == '\r'
+#endif
+          ) {
+        break;
+      }
+      if (state->growbufsiz == LINE_MAXLEN) {
+        discard = true;
+        break;
+      }
+
+      state->growbufsiz = (2 * state->growbufsiz < LINE_MAXLEN)
+        ? 2 * state->growbufsiz : LINE_MAXLEN;
+      state->growbuf = xrealloc(state->growbuf, 2 * state->growbufsiz);
```

Comment:
`state->growbufsiz` was already doubled on the line before, if needed.  This line should be using `state->growbufsiz`, not `2 * state->growbufsiz`.